fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/DeleteResourceServiceImpl.java

```
@@ -0,0 +1,122 @@
+ *
+ * @author dbernstein
+ */
+public class DeleteResourceServiceImpl implements DeleteResourceService {
+
+    private Logger LOGGER = LoggerFactory.getLogger(DeleteResourceService.class);
+
+    @Inject
+    private ContainmentIndex containmentIndex;
+
+    @Inject
+    private ResourceFactory resourceFactory;
+
+    @Inject
+    private DeleteResourceOperationFactory deleteResourceFactory;
+
+    @Inject
+    private PersistentStorageSessionManager psManager;
+
+    @Override
+    public void perform(final Transaction tx, final FedoraResource fedoraResource) {
+
+        final String fedoraResourceId = fedoraResource.getId();
+
+        try {
+            LOGGER.debug(format("deleting of %s", fedoraResourceId));
+            final PersistentStorageSession pSession = this.psManager.getSession(tx.getId());
+            deleteRecursivelyFromBottom(tx, pSession, fedoraResource);
+        } catch (final PersistentStorageException ex) {
+            throw new RepositoryRuntimeException(format("failed to delete resource %s", fedoraResourceId), ex);
+        }
+    }
+
+
+    private void deleteRecursivelyFromBottom(final Transaction tx, final PersistentStorageSession pSession,
+                                             final FedoraResource fedoraResource) throws PersistentStorageException {
+
+        final String fedoraId = fedoraResource.getId();
+
+        if (fedoraResource instanceof Container) {
+            final Stream<String> children = containmentIndex.getContainedBy(tx, fedoraResource);
+            children.forEach(childResourceId -> {
+                try {
+                    final FedoraResource res = resourceFactory.getResource(tx, childResourceId);
+                    deleteRecursivelyFromBottom(tx, pSession, res);
+                } catch (final PathNotFoundException ex) {
+                    LOGGER.warn(format("Path not found for %s: ignoring.", childResourceId));
```

I'm trying to think through what it would mean for the path to not be found. The index is out of sync? Or a concurrent operation already deleted the item? Trying to figure out the severity of this kind of failure.

Also, use the logger's formatting syntax, so `"Path not found for {}: ignoring.", childResourceId`

---

Yes - not sure myself.  We could throw a runtime exception here.